include/rocksdb/file_system.h

```diff
@@ -825,6 +860,38 @@ class FSRandomAccessFile {
     return IOStatus::NotSupported("InvalidateCache not supported.");
   }
 
+  // EXPERIMENTAL
+  // This API reads the requested data in FSReadRequest asynchronously. This is
+  // a asynchronous call, i.e it should return after submitting the request.
+  //
+  // When the read request is completed, callback function specified in cb
+  // should be called with arguments cb_arg and the result populated in
+  // FSReadResponse.
+  // cb_arg should be used by the callback to track the original request
+  // submitted.
+  //
+  // This API should also populate IOHandle which should be used by
+  // underlying FileSystem to store the context in order to distinguish the read
+  // requests at their side.
+  //
+  // req contains the request offset and size, but the result and status fields
+  // should be ignored. Instead, they should be passed in the FSReadResponse
+  // structure to the callback.
+  //
+  // Default implementation is to read the data synchronously.
+  virtual IOStatus ReadAsync(FSReadRequest* req, const IOOptions& opts,
+                             std::function<void(FSReadResponse*, void*)> cb,
+                             void* cb_arg, IOHandle* /*io_handle*/,
+                             IODebugContext* dbg) {
```

Comment:
The FSReadRequest and FSReadResponse should be a const &, not a pointer.  There is an effort to remove raw pointers from the public API

Reply:
Is there actually a need for an FSReadResponse?  The MultiRead uses the FSReadRequest for both input and output -- why can't this API do both?

Reply:
Yes, I think that's reasonable.